packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js

```diff
@@ -12,143 +11,42 @@ import Carousel from '../ProductImageCarousel';
-            variants
-        });
-
-        value = item
-            ? [...item.product.media_gallery_entries, ...media_gallery_entries]
-            : media_gallery_entries;
-    }
-
-    const key = value.reduce((fullKey, entry) => {
-        return `${fullKey},${entry.file}`;
-    }, '');
-
-    return { key, value };
-};
-
 const ProductFullDetail = props => {
-    const [{ isAddingItem }, { addItemToCart }] = useCartContext();
-
-    // Props.
     const { product } = props;
 
-    // State.
-    const [quantity, setQuantity] = useState(INITIAL_QUANTITY);
-    const [optionSelections, setOptionSelections] = useState(
-        INITIAL_OPTION_SELECTIONS
-    );
-    const derivedOptionCodes = deriveOptionCodesFromProduct(product);
-    const [optionCodes] = useState(derivedOptionCodes);
-
-    // Members.
-    const { amount: productPrice } = product.price.regularPrice;
-    const classes = mergeClasses(defaultClasses, props.classes);
-    const isMissingOptions = getIsMissingOptions(product, optionSelections);
-    const mediaGalleryEntries = getMediaGalleryEntries(
-        product,
-        optionCodes,
-        optionSelections
-    );
+    const talonProps = useProductFullDetail({
+        product
+    });
 
-    // Event handlers.
-    const handleAddToCart = useCallback(() => {
-        const payload = {
-            item: product,
-            productType: product.__typename,
-            quantity
-        };
+    const {
+        handleAddToCart,
+        handleSelectionChange,
+        handleSetQuantity,
+        isAddToCartDisabled,
+        mediaGalleryEntries,
+        productDetails,
+        quantity
+    } = talonProps;
 
-        if (isProductConfigurable(product)) {
-            appendOptionsToPayload(payload, optionSelections, optionCodes);
-        }
-
-        addItemToCart(payload);
-    }, [addItemToCart, optionCodes, optionSelections, product, quantity]);
-
-    const handleSelectionChange = useCallback(
-        (optionId, selection) => {
-            // We must create a new Map here so that React knows that the value
-            // of optionSelections has changed.
-            const newOptionSelections = new Map([...optionSelections]);
-            newOptionSelections.set(optionId, Array.from(selection).pop());
-            setOptionSelections(newOptionSelections);
-        },
-        [optionSelections]
-    );
+    const classes = mergeClasses(defaultClasses, props.classes);
 
     return (
         <Form className={classes.root}>
             <section className={classes.title}>
-                <h1 className={classes.productName}>{product.name}</h1>
+                <h1 className={classes.productName}>{productDetails.name}</h1>
                 <p className={classes.productPrice}>
                     <Price
-                        currencyCode={productPrice.currency}
-                        value={productPrice.value}
+                        currencyCode={productDetails.price.currency}
```

Comment:
Is there any way this could result in an error being thrown?

Reply:
https://github.com/magento/pwa-studio/pull/1814/files#diff-dd082926e212e0be673b35eb01025b3aL101

`productPrice` was previously just `amount` which is an object containing `currency` and `value` props. The normalization done in the talon is equivalent with only a rename of `productPrice` to `productDetails.price`.

Technically the old way (and the new method) can both throw if any of the values are undefined in the chain. Since we don't yet have option chaining I can leave these to some default.